src/storage/compressed_tile.cpp

```diff
@@ -0,0 +1,301 @@
+              j, old_column_offset, old_column_type, is_inlined);
+          Tile::SetValueFast(old_value, j, new_column_offset, old_column_type,
+                             is_inlined);
+        }
+
+      } else {
+        auto new_column_type = schema.GetColumn(i).GetType();
+
+        for (oid_t j = 0; j < num_tuple_slots; j++) {
+          Tile::SetValueFast(new_columns[i][j], j, new_column_offset,
+                             new_column_type, is_inlined);
+        }
+      }
+    }
+
+    is_compressed = true;
+  }
+}
+
+std::vector<type::Value> CompressedTile::CompressColumn(
+    Tile *tile, oid_t column_id,
+    type::Type::TypeId compression_type = type::Type::TINYINT) {
+  oid_t num_tuples = tile->GetAllocatedTupleCount();
+  auto tile_schema = tile->GetSchema();
+  bool is_inlined = tile_schema->IsInlined(column_id);
+  size_t column_offset = tile_schema->GetOffset(column_id);
+  auto column_info = tile_schema->GetColumn(column_id);
+  auto column_type = column_info.GetType();
+  std::vector<type::Value> column_values(num_tuples);
+
+  for (oid_t i = 0; i < num_tuples; i++) {
+    column_values[i] =
+        tile->GetValueFast(i, column_offset, column_type, is_inlined);
+  }
+
+  std::vector<type::Value> actual_values(column_values);
+
+  std::sort(column_values.begin(), column_values.end(), CompareLessThanBool);
+
+  type::Value median = column_values[num_tuples / 2];
+  std::vector<type::Value> modified_values(num_tuples);
+
+  while (true) {
+    try {
+      type::Value min_diff =
+          column_values[0].Subtract(median).CastAs(compression_type);
+      type::Value max_diff =
+          column_values[num_tuples - 1].Subtract(median).CastAs(
+              compression_type);
+      for (oid_t k = 0; k < num_tuples; k++) {
+        modified_values[k] =
+            actual_values[k].Subtract(median).CastAs(compression_type);
+      }
+
+      LOG_TRACE("Base value selected = %s", median.ToString().c_str());
+      break;
+    } catch (Exception &e) {
+      if (column_type == (compression_type + 1)) {
+        modified_values.resize(0);
+        break;
+      }
+      compression_type = static_cast<type::Type::TypeId>(
+          static_cast<int>(compression_type) + 1);
+    }
+  }
+  return modified_values;
+}
+
+void CompressedTile::InsertTuple(const oid_t tuple_offset, Tuple *tuple) {
+  if (IsCompressed()) {
+    LOG_TRACE("Peloton does not support insert into compressed tiles.");
+    PL_ASSERT(false);
+  } else {
+    Tile::InsertTuple(tuple_offset, tuple);
+  }
+}
+
+type::Value CompressedTile::GetValue(const oid_t tuple_offset,
+                                     const oid_t column_id) {
+  type::Value deserizedValue = Tile::GetValue(tuple_offset, column_id);
+
+  if (!IsCompressed()) {
+    return deserizedValue;
+  }
+
+  if (GetCompressedType(column_id) == type::Type::INVALID) {
+    return deserizedValue;
+  }
+
+  type::Type::TypeId compressed_type = GetCompressedType(column_id);
```

Comment:
It seems that a use in `PL_ASSERT` is not considered by the compiler in the given vagrant box, because the compilation using just `make` (rather than the tests) cannot run through.

> error: unused variable ‘compressed_type’